### PR TITLE
Client refactor and Explanations

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?
We refactor client-nodejs to follow the new architecture set out by client-java in 79634799c35a6de3085eed96386a5e059207bfea, and then following https://github.com/graknlabs/grakn/pull/6271 and the corresponding protocol change in https://github.com/graknlabs/protocol/pull/131 we implement Explanations, Explainable concept maps, and the explain() query API, which allows users to stream Explanations on demand **note: `explain` query or transaction option must be set to `true`**

## What are the changes implemented in this PR?
* completely overhaul client to match the architecture of client-java
* Implement `Explanation` objects, and extend `ConceptMap` to contain `Explainables`
* Add the `QueryManager.explain(Explainable)` API to retrieve all direct explanations (1-rule layer)

### CODE REVIEW

File: Grakn.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import {GraknClient as GC} from "./api/GraknClient";
```
Comment (alexjpwalker) [inline Grakn.ts:20]: Why do we alias to GC?
```suggestion
import {GraknClient} from "./api/GraknClient";
```

File: api/GraknOptions.ts [hunk 2]
```diff
@@ -17,12 +17,25 @@
...
+namespace Opts {
+    export class Core {
+        infer? : boolean;
+        traceInference? : boolean;
+        explain? : boolean;
+        parallel? : boolean;
+        batchSize? : number;
+        prefetch? : boolean;
+        sessionIdleTimeoutMillis? : number;
+        schemaLockAcquireTimeoutMillis? : number;
```
Comment (alexjpwalker) [inline api/GraknOptions.ts:31]: I don't think it's worth cleaning them up now, but it seems like your IDE may be using weird formatting defaults - inserting a space between ? and :, for instance.

File: api/concept/ConceptManager.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    putEntityType(label: string): Promise<EntityType>;
+
+    getRelationType(label: string): Promise<RelationType>;
+
+    putRelationType(label: string): Promise<RelationType>;
+
+    getAttributeType(label: string): Promise<AttributeType | null>;
+
+    putAttributeType(label: string, valueType: AttributeType.ValueType): Promise<AttributeType | null>;
```
Comment (alexjpwalker) [inline api/concept/ConceptManager.ts:51]: We can cut out these `| null` things, they aren't needed without `strictNullChecks` compiler option of `tsc`

File: api/concept/thing/Attribute.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    export interface DateTime extends Attribute<Date> {
+
+        asRemote(transaction: GraknTransaction): Attribute.RemoteDateTime;
+
+        getType(): AttributeType.DateTime;
+
+    }
+
+    export interface RemoteDateTime extends Attribute.Remote<Date>, Attribute.DateTime {
```
Comment (alexjpwalker) [inline api/concept/thing/Attribute.ts:130]: This isn't actually aligned with client-java - it should be `Attribute.Remote.DateTime`. We should either align fully with client-java, or revert to the most natural TypeScript name which would be `RemoteDateTimeAttribute`.

File: api/database/Database.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+export namespace Database {
+
+    export interface Cluster extends Database {
+
+        replicas(): Replica[];
+
+        primaryReplica(): Replica;
+
+        preferredSecondary(): Replica;
```
Comment (alexjpwalker) [inline api/database/Database.ts:39]: This is now called `preferredReplica` in client-java
```suggestion
        preferredReplica(): Replica;
```

File: api/logic/Explanation.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+import {Rule} from "./Rule";
+import {ConceptMap} from "../answer/ConceptMap";
+
+export interface Explanation {
+
+    rule(): Rule;
+
+    whenAnswer(): ConceptMap;
```
Comment (alexjpwalker) [inline api/logic/Explanation.ts:27]: Rename to `condition`

File: api/logic/Explanation.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    thenAnswer(): ConceptMap;
```
Comment (alexjpwalker) [inline api/logic/Explanation.ts:29]: Rename to `conclusion`

File: cluster/ClusterSession.ts [hunk 8]
```diff
@@ -77,54 +72,54 @@ export class SessionClusterRPC implements GraknClient.Session {
         return this._options;
     }
 
-    close(): Promise<void> {
+    async close(): Promise<void> {
```
Comment (alexjpwalker) [inline cluster/ClusterSession.ts:75]: We shouldn't use `async` here as there's no `await`
```suggestion
    close(): Promise<void> {
```

File: common/rpc/RequestBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,663 @@
...
+    AttributeType as AttributeTypeProto,
+    ConceptManager as ConceptMgrProto,
+    EntityType as EntityTypeProto,
+    Relation as RelationProto,
+    RelationType as RelationTypeProto,
+    RoleType as RoleTypeProto,
+    Thing as ThingProto,
+    ThingType as ThingTypeProto,
+    Type as TypeProto
+} from "grakn-protocol/common/concept_pb";
```
Comment (alexjpwalker) [inline common/rpc/RequestBuilder.ts:38]: While this probably works fine, it's not what we're doing in client-java (or Python) and may make the code a little harder to port in the future. We should refactor it at some point using:
```ts
import * as ConceptProto from "grakn-protocol/common/concept_pb";
```

File: core/CoreTransaction.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        return this._type;
+    }
+
+    public isOpen(): boolean {
+        return this._bidirectionalStream.isOpen();
+    }
+
+    public async rpcExecute(request: Transaction.Req, batch?: boolean): Promise<Transaction.Res> {
+        if (!this.isOpen()) throw new GraknClientError(TRANSACTION_CLOSED);
+        const useBatch = batch ? batch : true;
```
Comment (alexjpwalker) [inline core/CoreTransaction.ts:108]: This is a bug, `useBatch` will ALWAYS be `true`.
```suggestion
        const useBatch = batch !== false;
```

File: logic/ExplanationImpl.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+import {Rule} from "../api/logic/Rule";
+import {RuleImpl} from "./RuleImpl";
+import {ConceptMapImpl} from "../concept/answer/ConceptMapImpl";
+import {Explanation as ExplanationProto} from "grakn-protocol/common/logic_pb";
+
+export class ExplanationImpl implements Explanation {
+
+    private _rule: Rule;
+    private _whenAnswer: ConceptMap;
+    private _thenAnswer: ConceptMap;
```
Comment (alexjpwalker) [inline logic/ExplanationImpl.ts:31]: `_condition, _conclusion`

File: stream/BidirectionalStream.ts [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+import ResponseQueue = ResponseCollector.ResponseQueue;
+import TRANSACTION_CLOSED = ErrorMessage.Client.TRANSACTION_CLOSED;
+import MISSING_RESPONSE = ErrorMessage.Client.MISSING_RESPONSE;
+
+
+export class BidirectionalStream {
+
+    private readonly _requestTransmitter: RequestTransmitter;
+    private readonly _dispatcher: BatchDispatcher;
+    private readonly _responseCollector: ResponseCollector<unknown>;
```
Comment (alexjpwalker) [inline stream/BidirectionalStream.ts:40]: Not urgent to fix but this should be `ResponseCollector<TransactionProto.Transaction.Res>`

File: test/behaviour/connection/ConnectionStepsBase.ts [hunk 13]
```diff
@@ -57,6 +57,7 @@ Before(async () => {
 
 After(async () => {
     for (const session of sessions) {
+        console.log("closing sessions");
```
Comment (alexjpwalker) [inline test/behaviour/connection/ConnectionStepsBase.ts:60]: Do we still need this?

**Review by alexjpwalker — CHANGES_REQUESTED** [review]
Requested some changes

File: stream/ResponsePartIterator.ts [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+import {RequestBuilder} from "../common/rpc/RequestBuilder";
+import {ErrorMessage} from "../common/errors/ErrorMessage";
+import {GraknClientError} from "../common/errors/GraknClientError";
+import {Transaction} from "grakn-protocol/common/transaction_pb";
+import ResCase = Transaction.ResPart.ResCase;
+import MISSING_RESPONSE = ErrorMessage.Client.MISSING_RESPONSE;
+import UNKNOWN_STREAM_STATE = ErrorMessage.Client.UNKNOWN_STREAM_STATE;
+
+
+export class ResponsePartIterator implements AsyncIterable<Transaction.ResPart> {
```
[the remark this answers is not in the document]
Comment (alexjpwalker) [reply stream/ResponsePartIterator.ts:32]: This is now done in the Java and Python clients

File: common/rpc/RequestBuilder.ts (hunk 9 above)
Comment (flyingsilverfin) [reply common/rpc/RequestBuilder.ts:38]: decided to keep this, as its currently self-consistent in the node client

**Review by alexjpwalker — APPROVED** [review]

**Pull request merged**